WordPress/wordpress-playground
# Blueprints: Throw on failure

## Description

Ensures the Blueprint steps throw an error on failure. For now, the error is merely reported in the console. In the future, there may be a nice UI for this.

This PR is an addition to https://github.com/WordPress/wordpress-playground/pull/605 as it prepares the Blueprint steps for the changes in the Blueprint compilation engine.

Detecting a runPHP failure required adjusting the php_wasm.c code to return the correct exit code. The previous method of inferring the exit code was replaced by simply returning `EG(exit_status)` which is populated by the zend engine.

Furthermore, this PR ships additional unit tests for some Blueprint steps to ensure they indeed throw as expected.

## Changes description

### `throwOnError` option

`BasePHP.run()` now accepts a `throwOnError` option that throws an error whenever PHP returns an exit code different than `0`:

```js
// This will throw a JavaScript exception 
await php.run({
	throwOnError: true,
	code: '<?php no_such_function()'
});
```

This happens in the following cases:

 * syntax error
 * undefined function call
 * fatal error
 * calling `exit(1)`
 * uncaught PHP exception
 * ...probably some more

This option is set to `true` by all the Blueprint steps.

## Remaining work

- [x] Rebuild all the PHP.wasm versions

## Testing instructions

This PR comes with test coverage so confirm all the tests pass in CI.

Also, apply this PR locally and visit the following URL:

http://localhost:5400/website-server/#{%22steps%22:[{%22step%22:%22runPHP%22,%20%22code%22:%22%3C?php%20no_such_fn();%22}]}

It should reveal useful error information in the console.
